File: src/Watch.zig
const std = @import("std");
const df = @import("ImportC.zig").df;
const root = @import("root.zig");
const k = @import("Classes.zig").CLASS;
const Window = @import("Window.zig");
const q = @import("Message.zig");
const colors = @import("Colors.zig");

var tick:usize = 0;
const hands = [_][:0]const u8{" \xC0 ", " \xDA ", " \xBF ", " \xD9 "};
const bo = "\xCD";

pub fn WatchIconProc(win:*Window, msg: df.MESSAGE, params:q.Params) bool {
    switch (msg) {
        df.CREATE_WINDOW => {
            tick = 0;
            const rtn = root.DefaultWndProc(win, msg, params);
            _ = win.sendMessage(df.CAPTURE_MOUSE, .{.capture=.{false, null}});
            _ = win.sendMessage(df.HIDE_MOUSE, q.none);
            _ = win.sendMessage(df.CAPTURE_KEYBOARD, .{.capture=.{false, null}});
            _ = win.sendMessage(df.CAPTURE_CLOCK, q.none);
            return rtn;
            },
        df.CLOCKTICK => {
            tick = tick + 1;
            tick = tick & 3; // the same as tick % 4 for positive number
            if (win.PrevClock) |clock| {
                _ = clock.sendMessage(msg, params);
            } else { // could clock be null ?
                _ = q.SendMessage(null, msg, params);
            }
            // (fall through and paint)
            colors.SetStandardColor(win);
            win.writeline(hands[tick], 1, 1, false);
            return true;
        },
        df.PAINT => {
            colors.SetStandardColor(win);
            win.writeline(hands[tick], 1, 1, false);
            return true;
        },
        df.BORDER => {
            const rtn = root.DefaultWndProc(win, msg, params);
            win.writeline(bo, 2, 0, false);
            return rtn;
        },
        df.MOUSE_MOVED => {
            _ = win.sendMessage(df.HIDE_WINDOW, q.none);
            _ = win.sendMessage(df.MOVE, params);
            _ = win.sendMessage(df.SHOW_WINDOW, q.none);
            return true;
        },
        df.CLOSE_WINDOW => {
            _ = win.sendMessage(df.RELEASE_CLOCK, q.none);
            _ = win.sendMessage(df.RELEASE_MOUSE, .{.capture=.{false, null}});
            _ = win.sendMessage(df.RELEASE_KEYBOARD, .{.capture=.{false, null}});
            _ = win.sendMessage(df.SHOW_MOUSE, q.none);
        },
        else => {
        }
    }

    return root.DefaultWndProc(win, msg, params);
}

pub fn WatchIcon() *Window {
    var mx:usize = 10;
    var my:usize = 10;
    _ = q.SendMessage(null, df.CURRENT_MOUSE_CURSOR, .{.cursor=.{&mx, &my}});
    const win = Window.create(
                    k.BOX,
                    null,
                    mx, my, 3, 5,
                    null, null,
                    WatchIconProc,
                    df.VISIBLE | df.HASBORDER | df.SHADOW | df.SAVESELF, .{});
    return win;
}

File: watch.c
/* ----------- watch.c ----------- */

#include "dflat.h"

int WatchIconProc(WINDOW wnd, MESSAGE msg, PARAM p1, PARAM p2)
{
    int rtn, i;
	static int tick = 0;
	static const uint32_t hands[] = {
		0x2514, 0x250C, 0x2510, 0x2518
	};
    switch (msg)    {
        case CREATE_WINDOW:
			tick = 0;
            rtn = DefaultWndProc(wnd, msg, p1, p2);
            SendMessage(wnd, CAPTURE_MOUSE, 0, 0);
            SendMessage(wnd, HIDE_MOUSE, 0, 0);
            SendMessage(wnd, CAPTURE_KEYBOARD, 0, 0);
            SendMessage(wnd, CAPTURE_CLOCK, 0, 0);
            return rtn;
		case CLOCKTICK:
			++tick;
			tick &= 3;
			SendMessage(wnd->PrevClock, msg, p1, p2);
			/* (fall through and paint) */
        case PAINT:
            SetStandardColor(wnd);
            wputch(wnd, ' ', 1, 1);
            wputch(wnd, hands[tick], 2, 1);
            wputch(wnd, ' ', 3, 1);
            return TRUE;
        case BORDER:
            rtn = DefaultWndProc(wnd, msg, p1, p2);
            wputch(wnd, 0x2550, 2, 0);
            return rtn;
        case MOUSE_MOVED:
            SendMessage(wnd, HIDE_WINDOW, 0, 0);
            SendMessage(wnd, MOVE, p1, p2);
            SendMessage(wnd, SHOW_WINDOW, 0, 0);
            return TRUE;
        case CLOSE_WINDOW:
            SendMessage(wnd, RELEASE_CLOCK, 0, 0);
            SendMessage(wnd, RELEASE_MOUSE, 0, 0);
            SendMessage(wnd, RELEASE_KEYBOARD, 0, 0);
            SendMessage(wnd, SHOW_MOUSE, 0, 0);
            break;
        default:
            break;
    }
    return DefaultWndProc(wnd, msg, p1, p2);
}

WINDOW WatchIcon(void)
{
    int mx, my;
    WINDOW wnd;
    SendMessage(NULL, CURRENT_MOUSE_CURSOR,
                        (PARAM) &mx, (PARAM) &my);
    wnd = CreateWindow(
                    BOX,
                    NULL,
                    mx, my, 3, 5,
                    NULL,NULL,
                    WatchIconProc,
                    VISIBLE | HASBORDER | SHADOW | SAVESELF);
    return wnd;
}
